### ISSUE TITLE
[ppc64le] RuntimeDyldELF/TOC problem (segfaults)

### ISSUE THREAD

**Issue by abebeos** [numba/numba#6606]
(affects/blocks #4026) - (draft)

## Summary

A missing `nop` after a branch-instruction results in RuntimeDyldELF overwriting an instruction with the TOC restoration code. 

See bug-identification by `obilaniu` within https://github.com/numba/numba/issues/4026#issuecomment-487296000

This bug disappeared from mainline numba, because `reloc_mode` was changed to `pic`.

## Reproduction

Setup `llvm-10.0` `llvmlite-v0.35.0` `numba-0.52.0`. This test should work fine:

```sh
# within numba/
./runtests.py -vvv numba.tests.npyufunc.test_parallel_ufunc_issues.TestParGUfuncIssues
```

then patch numba to use `reloc_mode=static` for `ppc` and repeat the test, which should fail with a segfault:

```diff
diff --git a/numba/core/codegen.py b/numba/core/codegen.py
index 0efff305d..e8cd1df30 100644
--- a/numba/core/codegen.py
+++ b/numba/core/codegen.py
@@ -1297,7 +1297,7 @@ class JITCPUCodegen(BaseCPUCodegen):
         if arch.startswith('x86'): # one of x86 or x86_64
             reloc_model = 'static'
         elif arch.startswith('ppc'):
-            reloc_model = 'pic'
+            reloc_model = 'static'
         else:
             reloc_model = 'default'
         options['reloc'] = reloc_model
```

## Suggested Fix

The fix was identified during the work-in to the compiler-side of the problem. Whilst looking for similar work on the relevant (pointed out by `nemanjai` within https://github.com/numba/numba/issues/4026#issuecomment-657941602) `PPCISelLowering.cpp / allsShareTOCBase()`, the comments on the following commit indicated its relevance for the problem:

https://reviews.llvm.org/rG2812c1515627904e31605bbd4f25a887a1f8eb12

```diff
diff --git a/numba/core/codegen.py b/numba/core/codegen.py
index 0efff305d..e8cd1df30 100644
--- a/numba/core/codegen.py
+++ b/numba/core/codegen.py
@@ -1297,7 +1297,7 @@ class JITCPUCodegen(BaseCPUCodegen):
         if arch.startswith('x86'): # one of x86 or x86_64
             reloc_model = 'static'
         elif arch.startswith('ppc'):
-            reloc_model = 'pic'
+            reloc_model = 'static'
         else:
             reloc_model = 'default'
         options['reloc'] = reloc_model
(base) root@61ebdd0b8bae:~# cd mine
(base) root@61ebdd0b8bae:~/mine# cat weak_GV_add_TOC_nop_by_stefanp.patch
diff --git a/llvm/lib/Target/PowerPC/PPCISelLowering.cpp b/llvm/lib/Target/PowerPC/PPCISelLowering.cpp
index ca1649fae..23124708d 100644
--- a/llvm/lib/Target/PowerPC/PPCISelLowering.cpp
+++ b/llvm/lib/Target/PowerPC/PPCISelLowering.cpp
@@ -4558,7 +4558,15 @@ callsShareTOCBase(const Function *Caller, SDValue Callee,
    if (!G)
      return false;

-   const GlobalValue *GV = G->getGlobal();
+  const GlobalValue *GV = G->getGlobal();
+
+  // If the GV is not a strong definition then we need to assume it can be
+  // replaced by another function at link time. The function that replaces
+  // it may not share the same TOC as the caller since the callee may be
+  // replaced by a PC Relative version of the same function.
+  if (!GV->isStrongDefinitionForLinker())
+    return false;
+
   // The medium and large code models are expected to provide a sufficiently
   // large TOC to provide all data addressing needs of a module with a
   // single TOC. Since each module will be addressed with a single TOC then we
@@ -4567,13 +4575,6 @@ callsShareTOCBase(const Function *Caller, SDValue Callee,
       CodeModel::Large == TM.getCodeModel())
     return TM.shouldAssumeDSOLocal(*Caller->getParent(), GV);

-  // Otherwise we need to ensure callee and caller are in the same section,
-  // since the linker may allocate multiple TOCs, and we don't know which
-  // sections will belong to the same TOC base.
-
-  if (!GV->isStrongDefinitionForLinker())
-    return false;
-
   // Any explicitly-specified sections and section prefixes must also match.
   // Also, if we're using -ffunction-sections, then each function is always in
   // a different section (the same is true for COMDAT functions).
```

## Validation

(todo, numba repos seem to not verify for ppc)

```sh
# on ppc64le
# You need to have anaconda 
conda create -n ppc-segfault python=3.6 numpy scipy jinja2 cffi
conda activate ppc-segfault

# install llvm, llvmlite, numba from source
# apply the 2 patches
# runt the tests
```

**Comment by abebeos** [comment]
> We at IBM are concerned that there is a latent bug in RuntimeDyldELF for PPC64LE, but apparently there is no reproducer.

source: https://github.com/numba/numba/issues/4026#issuecomment-705771587

I can say with near 100% confidence: no latent bugs in the code (re this issue). The compiler-side code I reviewed (PPCISelLowering.cpp) is very clean, with the right amount of in-source documentation.

The work/reviews on e.g. https://reviews.llvm.org/D81126?id=274770 can be easily followed.

With a little bit effort for clarifying RuntimeDyldELF (only the code that directly affects ppc), all should be good. The clarity of RuntimeDyldELF should become similar that of PPCISelLowering.cpp, even if the future direction will be JITLink

**Comment by stuartarchibald** [comment]
@abebeos Thanks for the patch(es) and for fixing this! As promised [here](https://github.com/numba/numba/pull/6300#issuecomment-753953184) I raised this issue at yesterday's development meeting. Outcome was that we're going to hopefully roll these fixes into the 0.54 release of Numba along with attempting an upgrade to LLVM 11.

**Comment by abebeos** [comment]
All good, just 2 things:

* I'd still need a validation from your side (or from a 3rd party), just to be sure. It looks like @esc (https://github.com/numba/llvmlite/pull/684#issuecomment-755211233) would be doing this?
* this work is in context of of a bounty (bountysource), and bounties are not processed until the issues are closed. So possibly after validation, a merge could be done at least on a separate branch, in order to close this issue (and the blocked #4026).

I could push further packages to anaconda, to simplify your validation. Just let me know (currently out: https://anaconda.org/abebeos/llvmdev).

**Comment by esc** [comment]
@abebeos thank you again for submitting this! We appreciate your efforts to improve Numba. Regarding the timeline: the 0.54 release cycle will most likely begin in a few weeks from now (end of Jan/ beg. of Feb). At that point I can commit to reviewing and validating your patch. I should also note, ahead of time, that we may need this patch to be ported to LLVM 11.0 since we will be attempting an upgrade from LLVM 10 to LLVM 11 as part of the 0.54 release.

**Comment by abebeos** [comment]
>> this work is in context of of a bounty (bountysource), and bounties are not processed until the issues are closed.
>
> will most likely begin in a few weeks from now (end of Jan/ beg. of Feb)

Ouch!

I decided to apply the patch here via your patch-system, as I expected this to be the fastest way.

Trying now to bring this into 10.x, directly at llvm (https://reviews.llvm.org/D94183), possibly its faster. Moving it then into llvm11 should be a simple followup step.

**Comment by abebeos** [comment]
@esc, note that I've updated the Repro/Validate sections, it is far simpler now.

**Comment by esc** [comment]
@abebeos thank you, I will take it into consideration.

**Comment by esc** [comment]
Dear @abebeos,

Judging from your response of "Ouch!" I assume that you are disappointed about the proposed timeline for validating your proposed patches? If that is the case, I am sorry that the project schedule does to align with your expectations. 

Also, thank you for pre-compiling those packages but unfortunately they will be of little use for me. We will want to build and test the Numba stack (`llvmdev` -> `llvmlite` -> `numba`) across the supported versions of Python and Numpy on the corresponding supported architectures as part of our efforts in quality assurance. 

Thank you in advance for your patience and understanding in these matters.

-esc

**Comment by abebeos** [comment]
Nah, that "ouch" is mostly kind of "An automated CI would have validate minutes after check-in and would have made binary packages redundant",

Filed a dedicated issue here: #6629
